src/Features/Core/Portable/EditAndContinue/Remote/RemoteEditAndContinueServiceProxy.cs

```diff
@@ -0,0 +1,462 @@
+                    WellKnownServiceHubService.RemoteEditAndContinueService,
+                    callbackTarget: new StartEditSessionCallback(activeStatementProvider, debuggeeModuleMetadataProvider),
+                    cancellationToken).ConfigureAwait(false);
+
+                documentsToReanalyze = await connection.RunRemoteAsync<ImmutableArray<DocumentId>>(
+                    nameof(IRemoteEditAndContinueService.StartEditSessionAsync),
+                    solution: null,
+                    Array.Empty<object>(),
+                    cancellationToken).ConfigureAwait(false);
+
+                result = connection;
+            }
+
+            // clear all reported run mode diagnostics:
+            diagnosticService.Reanalyze(_workspace, documentIds: documentsToReanalyze);
+
+            return result;
+        }
+
+        private sealed class LocalConnection : IDisposable
+        {
+            public static readonly LocalConnection Instance = new LocalConnection();
+
+            public void Dispose()
+            {
+            }
+        }
+
+        private sealed class StartEditSessionCallback : IRemoteEditAndContinueService.IStartEditSessionCallback
+        {
+            private readonly ActiveStatementProvider _activeStatementProvider;
+            private readonly IDebuggeeModuleMetadataProvider _managedModuleInfoProvider;
+
+            public StartEditSessionCallback(ActiveStatementProvider activeStatementProvider, IDebuggeeModuleMetadataProvider debuggeeModuleMetadataProvider)
+            {
+                _activeStatementProvider = activeStatementProvider;
+                _managedModuleInfoProvider = debuggeeModuleMetadataProvider;
+            }
+
+            /// <summary>
+            /// Remote API.
+            /// </summary>
+            public async Task<ImmutableArray<ActiveStatementDebugInfo>> GetActiveStatementsAsync(CancellationToken cancellationToken)
+            {
+                try
+                {
+                    return await _activeStatementProvider(cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+                {
+                    return ImmutableArray<ActiveStatementDebugInfo>.Empty;
+                }
+            }
+
+            /// <summary>
+            /// Remote API.
+            /// </summary>
+            public async Task<(int errorCode, string? errorMessage)?> GetEncAvailabilityAsync(Guid mvid, CancellationToken cancellationToken)
+            {
+                try
+                {
+                    return await _managedModuleInfoProvider.GetEncAvailabilityAsync(mvid, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+                {
+                    // TODO: better error code?
+                    return (-1, e.Message);
```

Comment:
also, bettter message potentially (if this is going to be user visbile).  i.e. if we just pop up something saying "null ref at blah", that's not great.